src/ott/neural/models/base_solver.py

```diff
@@ -0,0 +1,579 @@
+        geom_xx,
+        geom_yy,
+        geom_xy,
+        fused_penalty=fused_penalty,
+        tau_a=tau_a,
+        tau_b=tau_b
+    )
+    return ot_solver(prob)
+
+  return match_pairs
+
+
+class BaseOTMatcher:
+  """Base class for mini-batch neural OT matching classes."""
+
+  def _resample_data(
+      self,
+      rng: jax.Array,
+      tmat: jnp.ndarray,
+      source_arrays: Tuple[jnp.ndarray, ...],
+      target_arrays: Tuple[jnp.ndarray, ...],
+  ) -> Tuple[jnp.ndarray, ...]:
+    """Resample a batch according to coupling `tmat`."""
+    tmat_flattened = tmat.flatten()
+    indices = jax.random.choice(rng, len(tmat_flattened), p=tmat_flattened, shape=[tmat.shape[0]])
+    indices_source = indices // tmat.shape[1]
+    indices_target = indices % tmat.shape[1]
+    return tuple(
+        b[indices_source] if b is not None else None for b in source_arrays
+    ), tuple(
+        b[indices_target] if b is not None else None for b in target_arrays
+    )
+
+  def _sample_conditional_indices_from_tmap(
+      self,
+      rng: jax.Array,
+      tmat: jnp.ndarray,
+      *,
+      k_samples_per_x: Union[int, jnp.ndarray],
+      source_arrays: Tuple[jnp.ndarray, ...],
+      target_arrays: Tuple[jnp.ndarray, ...],
+      source_is_balanced: bool,
+  ) -> Tuple[jnp.array, jnp.array]:
+    batch_size = tmat.shape[0]
+    left_marginals = tmat.sum(axis=1)
+    if not source_is_balanced:
+      rng, rng_2 = jax.random.split(rng, 2)
+      indices = jax.random.choice(
+          key=rng_2,
+          a=jnp.arange(len(left_marginals)),
+          p=left_marginals,
+          shape=(len(left_marginals),)
+      )
+    else:
+      indices = jnp.arange(batch_size)
+    tmat_adapted = tmat[indices]
+    indices_per_row = jax.vmap(
+        lambda row: jax.random.choice(
+            key=rng, a=jnp.arange(batch_size), p=row, shape=(k_samples_per_x,)
+        ),
+        in_axes=0,
+        out_axes=0,
+    )(
+        tmat_adapted
+    )
+
+    indices_source = jnp.repeat(indices, k_samples_per_x)
+    indices_target = jnp.reshape(
+        indices_per_row % tmat.shape[1], (batch_size * k_samples_per_x,)
+    )
+    return tuple(
+        jnp.reshape(b[indices_source], (k_samples_per_x, batch_size,
+                                        -1)) if b is not None else None
+        for b in source_arrays
+    ), tuple(
+        jnp.reshape(b[indices_target], (k_samples_per_x, batch_size,
+                                        -1)) if b is not None else None
+        for b in target_arrays
+    )
+
+
+class OTMatcherLinear(BaseOTMatcher):
+  """Class for mini-batch OT in neural optimal transport solvers.
+
+  Args:
+    ot_solver: OT solver to match samples from the source and the target
+      distribution as proposed in :cite:`tong:23`, :cite:`pooladian:23`.
+      If :obj:`None`, no matching will be performed as proposed in
+      :cite:`lipman:22`.
+  """
+
+  def __init__(
+      self,
+      ot_solver: was_solver.WassersteinSolver,
```

Comment:
Is this the correct typehint?  `Sinkhorn` is used in tests.

Reply:
We could also use `LRSinkhorn` https://github.com/ott-jax/ott/blob/851918218edb90d7f8068acf13fe905bdfa88ddf/src/ott/solvers/linear/sinkhorn_lr.py#L270. But you're right, now `Sinkhorn` is not a subclass of `WassersteinSolver` any more, and `LRSinkhorn` subclasses `Sinkhorn`, so `Sinkhorn` should be the correct type.